virtio-devices/src/vdpa.rs

```diff
@@ -0,0 +1,374 @@
+    GetBackendFeatures(vhost::Error),
+    #[error("Failed to get virtio features: {0}")]
+    GetFeatures(vhost::Error),
+    #[error("Failed to get the IOVA range: {0}")]
+    GetIovaRange(vhost::Error),
+    #[error("Failed to get queue size: {0}")]
+    GetVringNum(vhost::Error),
+    #[error("Invalid IOVA range")]
+    InvalidIovaRange,
+    #[error("Missing VIRTIO_F_ACCESS_PLATFORM feature")]
+    MissingAccessPlatformVirtioFeature,
+    #[error("Failed to reset owner: {0}")]
+    ResetOwner(vhost::Error),
+    #[error("Failed to set backend specific features: {0}")]
+    SetBackendFeatures(vhost::Error),
+    #[error("Failed to set eventfd notifying about a configuration change: {0}")]
+    SetConfigCall(vhost::Error),
+    #[error("Failed to set virtio features: {0}")]
+    SetFeatures(vhost::Error),
+    #[error("Failed to set memory table: {0}")]
+    SetMemTable(vhost::Error),
+    #[error("Failed to set owner: {0}")]
+    SetOwner(vhost::Error),
+    #[error("Failed to set virtio status: {0}")]
+    SetStatus(vhost::Error),
+    #[error("Failed to set vring address: {0}")]
+    SetVringAddr(vhost::Error),
+    #[error("Failed to set vring base: {0}")]
+    SetVringBase(vhost::Error),
+    #[error("Failed to set vring eventfd when buffer are used: {0}")]
+    SetVringCall(vhost::Error),
+    #[error("Failed to enable/disable vring: {0}")]
+    SetVringEnable(vhost::Error),
+    #[error("Failed to set vring eventfd when new descriptors are available: {0}")]
+    SetVringKick(vhost::Error),
+    #[error("Failed to set vring size: {0}")]
+    SetVringNum(vhost::Error),
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+pub struct Vdpa {
+    common: VirtioCommon,
+    id: String,
+    vhost: VhostKernVdpa<GuestMemoryAtomic<GuestMemoryMmap>>,
+    iova_range: VhostVdpaIovaRange,
+    enabled_num_queues: Option<usize>,
+    backend_features: u64,
+}
+
+impl Vdpa {
+    pub fn new(
+        id: String,
+        device_path: &str,
+        mem: GuestMemoryAtomic<GuestMemoryMmap>,
+        num_queues: u16,
+    ) -> Result<Self> {
+        let mut vhost = VhostKernVdpa::new(device_path, mem).map_err(Error::CreateVhostVdpa)?;
+        vhost.set_owner().map_err(Error::SetOwner)?;
+        let device_type = vhost.get_device_id().map_err(Error::GetDeviceId)?;
+        let queue_size = vhost.get_vring_num().map_err(Error::GetVringNum)?;
+        let avail_features = vhost.get_features().map_err(Error::GetFeatures)?;
+        let backend_features = vhost
+            .get_backend_features()
+            .map_err(Error::GetBackendFeatures)?;
+        vhost.set_backend_features_acked(backend_features);
+
+        // There's a bug in rust-vmm/vhost crate. Let's wait for next release
+        // to fix it. Below is the correct code once the bug will be fixed:
+        //
+        // let iova_range = vhost.get_iova_range().map_err(Error::GetIovaRange)?;
+        let iova_range = VhostVdpaIovaRange {
```

Comment:
For situations like this I would like us to have a bug open in our issue tracker which we can then use to track this.

Reply:
#3861